File: Portfolio/Portfolio.Blazor/Data/Entities/ContentRelation.cs
namespace Portfolio.Blazor.Data.Entities;

public sealed class ContentRelation
{
    public int Id { get; set; }
    public int RelationTypeId { get; set; }
    public string SubjectType { get; set; } = string.Empty;
    public int SubjectId { get; set; }
    public string ObjectType { get; set; } = string.Empty;
    public int ObjectId { get; set; }
    public string? Note { get; set; }
    public string? Context { get; set; }
    public string? Status { get; set; }
    public string? Visibility { get; set; }
    public DateTime? CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }

    public RelationType? RelationType { get; set; }
}

File: laravel/app/Models/ContentRelation.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\MorphTo;

class ContentRelation extends Model
{
    use HasFactory;

    protected $fillable = ['relation_type_id', 'subject_type', 'subject_id', 'object_type', 'object_id', 'note', 'context', 'status', 'visibility'];

    /**
     * @return BelongsTo<RelationType, $this>
     */
    public function relationType(): BelongsTo
    {
        return $this->belongsTo(RelationType::class);
    }

    public function subject(): MorphTo
    {
        return $this->morphTo();
    }

    public function object(): MorphTo
    {
        return $this->morphTo();
    }
}
